config/configmapprovider/provider.go

```diff
@@ -26,36 +26,72 @@ type Provider interface {
 	// Retrieve goes to the configuration source and retrieves the selected data which
 	// contains the value to be injected in the configuration and the corresponding watcher that
 	// will be used to monitor for updates of the retrieved value.
+	// Should never be called concurrently with itself.
 	Retrieve(ctx context.Context) (Retrieved, error)
 
 	// Close signals that the configuration for which it was used to retrieve values is no longer in use
 	// and the object should close and release any watchers that it may have created.
 	// This method must be called when the service ends, either in case of success or error.
+	// Should never be called concurrently with itself.
 	Close(ctx context.Context) error
+
+	// Note that Retrieve() and Close() should never be called concurrently. Close()
+	// should be called only after Retrieve() returns.
 }
 
 // Retrieved holds the result of a call to the Retrieve method of a Session object.
 type Retrieved interface {
 	// Get returns the Map.
+	// Should never be called concurrently with itself.
 	Get() *config.Map
 }
 
 // WatchableRetrieved is an extension for Retrieved that if implemented,
 // the Retrieved value supports monitoring for updates.
+//
+// The typical usage is the following:
+//
+//	goroutine 1                                        | go routine 2
+//	                                                   |
+//	r := mapProvider.Retrieve()                        |
+//	r.Get()                                            |
+//	r.WatchForUpdate() blocks until there are updates  |
+//	...                                                |
+//	r.Get()                                            |
+//	r.WatchForUpdate()                                 | Close() can be called anytime after Retrieve() returns
```

Comment:
This is not what we have in the collector right now. We call `r.Close` after every`r.WatchForUpdate()` then `mapProvider.Retrieve()` again.